File: samples/cpp/SampleApp/src/CBL/CBLHandler.cpp
/*
 * Copyright 2017-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

#include "SampleApp/CBL/CBLHandler.h"

#include "SampleApp/ApplicationContext.h"

// C++ Standard Library
#include <sstream>

// Guidelines Support Library
#define GSL_THROW_ON_CONTRACT_VIOLATION
#include <gsl/contracts.h>

namespace sampleApp {
namespace cbl {

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
//  CBLHandler
//
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

CBLHandler::CBLHandler(std::weak_ptr<Activity> activity, std::weak_ptr<logger::LoggerHandler> loggerHandler)
    : m_activity{std::move(activity)}, m_loggerHandler{std::move(loggerHandler)} {
    // Expects((m_activity != nullptr) && (m_loggerHandler != nullptr));
    setupUI();
}

std::weak_ptr<Activity> CBLHandler::getActivity() { return m_activity; }

std::weak_ptr<logger::LoggerHandler> CBLHandler::getLoggerHandler() { return m_loggerHandler; }

// aace::cbl::CBL interface

void CBLHandler::cblStateChanged(CBLState state, CBLStateChangedReason reason, const std::string &url, const std::string &code) {
    std::stringstream ss;
    ss << "cblStateChanged:state=" << state << ",reason=" << reason << ",url=" << url << ",code=" << code;
    log(logger::LoggerHandler::Level::INFO, ss.str());
    auto activity = m_activity.lock();
    if (!activity) {
        return;
    }
    switch (state) {
        case CBLState::CODE_PAIR_RECEIVED: {
            activity->runOnUIThread([=]() {
                if (auto card = activity->findViewById("id:card").lock()) {
                    auto string = "{\"code\":\"" + code + "\",\"url\":\"" + url + "\"}";
                    card->set(string, View::Type::CBLCode);
                }
            });
            auto command = m_applicationContext->getBrowserCommand();
            if (!command.empty()) {
                m_applicationContext->executeCommand((command + ' ' + url + "?cbl-code=" + code).c_str());
            }
            break;
        }
        case CBLState::STARTING:
            m_busy = true;
            break;
        case CBLState::STOPPING:
            m_busy = false;
            switch (reason) {
                case CBLStateChangedReason::CODE_PAIR_EXPIRED:
                    activity->runOnUIThread([=]() {
                        if (auto card = activity->findViewById("id:card").lock()) {
                            auto string = "{\"message\":\"The code has expired. Retry to generate a new code.\"}";
                            card->set(string, View::Type::CBLCodeExpired);
                        }
                    });
                    break;
                default:
                    break;
            }
            break;
        default:
            break;
    }
}

void CBLHandler::clearRefreshToken() {
    Ensures(m_applicationContext != nullptr);
    return m_applicationContext->clearRefreshToken();
}

std::string CBLHandler::getRefreshToken() {
    Ensures(m_applicationContext != nullptr);
    return m_applicationContext->getRefreshToken();
}

void CBLHandler::setRefreshToken(const std::string &refreshToken) {
    // IMPORTANT: YOUR PRODUCT IS RESPONSIBLE FOR STORING THE REFRESH TOKEN SECURELY.
    // FOR SECURITY REASONS, AUTHENTICATION IS NOT PRESERVED IN THE C++ SAMPLE APP.
    Ensures(m_applicationContext != nullptr);
    m_applicationContext->setRefreshToken(refreshToken);
}

// private

void CBLHandler::log(logger::LoggerHandler::Level level, const std::string &message) {
    auto loggerHandler = m_loggerHandler.lock();
    if (!loggerHandler) {
        return;
    }
    loggerHandler->log(level, "CBLHandler", message);
}

void CBLHandler::setupUI() {
    auto activity = m_activity.lock();
    if (!activity) {
        return;
    }
    m_applicationContext = activity->getApplicationContext();
    m_console = activity->findViewById("id:console");

    // start
    activity->registerObserver(Event::onCBLStart, [=](const std::string &) {
        log(logger::LoggerHandler::Level::VERBOSE, "onCBLStart");
        if (m_applicationContext->hasRefreshToken()) {
            if (auto console = m_console.lock()) {
                console->printLine("You already have your refresh token");
            }
            return false;
        }
        if (m_busy) {
            if (auto console = m_console.lock()) {
                console->printLine("You already started CBL login flow");
            }
            return false;
        }
        if (auto console = m_console.lock()) {
            console->printLine("Starting CBL login flow...");
        }
        start();
        return true;
    });

    // cancel
    activity->registerObserver(Event::onCBLCancel, [=](const std::string &) {
        log(logger::LoggerHandler::Level::VERBOSE, "onCBLCancel");
        if (m_applicationContext->hasRefreshToken()) {
            if (auto console = m_console.lock()) {
                console->printLine("You already have your refresh token");
            }
            return false;
        }
        if (!m_busy) {
            if (auto console = m_console.lock()) {
                console->printLine("You have not started CBL login flow");
            }
            return false;
        }
        if (auto console = m_console.lock()) {
            console->printLine("Canceling CBL login flow...");
        }
        cancel();
        return true;
    });
}

} // namespace cbl
} // namespace sampleApp

File: samples/android/app/src/main/java/com/amazon/sampleapp/impl/CBL/CBLHandler.java
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amazon.sampleapp.impl.CBL;

import android.app.Activity;
import android.view.View;
import android.widget.ImageButton;
import android.widget.TextView;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;

import com.amazon.aace.cbl.CBL;
import com.amazon.aace.alexa.AlexaClient.AuthState;
import com.amazon.aace.alexa.AlexaClient.AuthError;
import com.amazon.sampleapp.R;
import com.amazon.sampleapp.impl.Logger.LoggerHandler;
import com.amazon.sampleapp.logView.LogRecyclerViewAdapter;
import com.amazon.sampleapp.impl.AlexaClient.AuthStateObserver;

import org.json.JSONException;
import org.json.JSONObject;

public class CBLHandler extends CBL implements AuthStateObserver {

    private static final String sTag = "CBL";

    private final LoggerHandler mLogger;
    private final Activity  mActivity;
    private final Handler mMainThreadHandler;
    private final SharedPreferences mPreferences;

    private View mLoginView, mCancelView, mLogoutView;

    public CBLHandler( Activity activity, LoggerHandler logger ) {
        mActivity = activity;
        mLogger = logger;
        mMainThreadHandler = new Handler(Looper.getMainLooper());
        mPreferences = activity.getSharedPreferences( activity.getString( R.string.preference_file_key ), Context.MODE_PRIVATE );
        setupUI();
    }

    @Override
    public void cblStateChanged( CBLState state, CBLStateChangedReason reason, String url, String code ) {
        mLogger.postInfo( sTag, String.format( "cblStateChanged:state=%s,reason=%s,url=%s code=%s", state.name(), reason.name(), url, code ) );

        switch ( state ) {
            case CODE_PAIR_RECEIVED:
                try {
                    JSONObject renderJSON = new JSONObject();
                    renderJSON.put( "verification_uri", url );
                    renderJSON.put( "user_code", code );
                    mLogger.postDisplayCard( renderJSON, LogRecyclerViewAdapter.CBL_CODE );
                } catch ( Exception e ) {
                    mLogger.postError( sTag, e.getMessage() );
                }
                break;
            case STARTING:
                showCancelButton();
                break;
            case STOPPING:
                switch ( reason ) {
                    case CODE_PAIR_EXPIRED:
                        showLoginButton();
                        try {
                            JSONObject renderJSON = new JSONObject();
                            String expiredMessage = "The code has expired. Retry to generate a new code.";
                            renderJSON.put( "message", expiredMessage );
                            mLogger.postDisplayCard( renderJSON, LogRecyclerViewAdapter.CBL_CODE_EXPIRED );
                        } catch ( JSONException e ) {
                            mLogger.postError( sTag, e.getMessage() );
                        }
                        break;
                    case AUTHORIZATION_EXPIRED:
                        try {
                            JSONObject renderJSON = new JSONObject();
                            String expiredMessage = "The token has expired. Log in again.";
                            renderJSON.put( "message", expiredMessage );
                            mLogger.postDisplayCard( renderJSON, LogRecyclerViewAdapter.CBL_CODE_EXPIRED );
                        } catch ( JSONException e ) {
                            mLogger.postError( sTag, e.getMessage() );
                        }
                        break;
                    case NONE:
                        // CBL stopped using cancel button
                        showLoginButton();
                    default:
                        break;
                }
                break;
            default:
                break;
        }
    }

    @Override
    public void clearRefreshToken() {
        SharedPreferences.Editor editor = mPreferences.edit();
        editor.putString( mActivity.getString( R.string.preference_refresh_token ), "" );
        editor.apply();
    }

    @Override
    public void setRefreshToken( String refreshToken ) {
        SharedPreferences.Editor editor = mPreferences.edit();
        editor.putString( mActivity.getString( R.string.preference_refresh_token ), refreshToken );
        editor.apply();
    }

    @Override
    public String getRefreshToken() {
        return mPreferences.getString( mActivity.getString( R.string.preference_refresh_token ), "" );
    }

    @Override
    public void setUserProfile( String name, String email ) {
        mLogger.postInfo( sTag, "User profile details updated." );
    }

    @Override
    public void onAuthStateChanged( final AuthState authState, final AuthError authError ) {
        mMainThreadHandler.post( new Runnable() {
            @Override
            public void run() {
                if ( authState == AuthState.REFRESHED ) {
                    showLogoutButton();
                } else {
                    showLoginButton();
                }
            }
        } );
    }

    private void setupUI() {
        mLoginView = mActivity.findViewById( R.id.login );
        mLoginView.findViewById( R.id.loginButton ).setOnClickListener( new View.OnClickListener() {
            @Override
            public void onClick( View v ) {
                mLogger.postInfo( sTag, "Starting CBL login flow..." );
                start();
            }
        } );

        mCancelView = mActivity.findViewById( R.id.cancel );
        mCancelView.findViewById( R.id.cancelButton ).setOnClickListener( new View.OnClickListener() {
            @Override
            public void onClick( View v ) {
                mLogger.postInfo( sTag, "Cancelling CBL login flow..." );
                cancel();
            }
        } );

        mLogoutView = mActivity.findViewById( R.id.logout );
        mLogoutView.findViewById( R.id.logoutButton ).setOnClickListener( new View.OnClickListener() {
            @Override
            public void onClick( View v ) {
                mLogger.postInfo( sTag, "Resetting CBL..." );
                reset();
            }
        } );
    }

    private void showLoginButton() {
        mMainThreadHandler.post( new Runnable() {
            @Override
            public void run() {
                mLoginView.setVisibility( View.VISIBLE );
                mCancelView.setVisibility( View.GONE );
                mLogoutView.setVisibility( View.GONE );
            }
        } );
    }

    private void showCancelButton() {
        mMainThreadHandler.post( new Runnable() {
            @Override
            public void run() {
                mLoginView.setVisibility( View.GONE ) ;
                mCancelView.setVisibility( View.VISIBLE );
                mLogoutView.setVisibility( View.GONE );
            }
        } );
    }

    private void showLogoutButton() {
        mMainThreadHandler.post( new Runnable() {
            @Override
            public void run() {
                mLoginView.setVisibility( View.GONE ) ;
                mCancelView.setVisibility( View.GONE );
                mLogoutView.setVisibility( View.VISIBLE );
            }
        } );
    }
}
